lib/rubygems/text.rb

```diff
@@ -49,37 +49,54 @@ def min3(a, b, c) # :nodoc:
     end
   end
 
-  # This code is based directly on the Text gem implementation
   # Returns a value representing the "cost" of transforming str1 into str2
+  # Vendored version of DidYouMean::Levenshtein.distance from the did_you_mean gem
+  # https://git.io/JJgZI
   def levenshtein_distance(str1, str2)
-    s = str1
-    t = str2
-    n = s.length
-    m = t.length
-
-    return m if (0 == n)
-    return n if (0 == m)
+    n = str1.length
+    m = str2.length
+    return m if n.zero?
+    return n if m.zero?
 
     d = (0..m).to_a
     x = nil
 
-    str1.each_char.each_with_index do |char1,i|
-      e = i + 1
+    # to avoid duplicating an enumerable object, create it outside of the loop
+    str2_codepoints = str2.codepoints
 
-      str2.each_char.each_with_index do |char2,j|
-        cost = (char1 == char2) ? 0 : 1
+    str1.each_codepoint.with_index(1) do |char1, i|
+      j = 0
+      while j < m
+        cost = (char1 == str2_codepoints[j]) ? 0 : 1
         x = min3(
-             d[j + 1] + 1, # insertion
-             e + 1,      # deletion
-             d[j] + cost # substitution
-           )
-        d[j] = e
-        e = x
+          d[j + 1] + 1, # insertion
+          i + 1,      # deletion
+          d[j] + cost # substitution
+        )
+        d[j] = i
+        i = x
+
+        j += 1
       end
-
       d[m] = x
     end
 
-    return x
+    x
   end
 end
+
+# Use the latest and greatest DidYouMean::Levenshtein if it is available
+begin
+  require 'did_you_mean/levenshtein'
+
+  module Gem::Text::LevenshteinExt
+    def levenshtein_distance(str1, str2)
+      DidYouMean::Levenshtein.distance(str1, str2)
+    end
+  end
+
+  Gem::Text.prepend(Gem::Text::LevenshteinExt)
+
+rescue LoadError
+  # ignored
+end
```

Comment:
Can you wrap this in `defined?(DidYouMean)`?
If the user did `--disable-did_you_mean` then RubyGems should not load it.

Also maybe only `defined?(DidYouMean::Levenshtein)` is enough (no `require`, no `rescue`) since did_you_mean.rb loads did_you_mean/levenshtein.rb eagerly.